test-harness/tests/test_master.py

```diff
@@ -134,3 +137,78 @@ def test_if_upstream_headers_are_correct(self,
                                              valid_user_header,
                                              path_fmt.format(prefix),
                                              )
+
+
+class TestSystemApiLeaderProxing():
+    def test_if_request_is_sent_to_the_current_mesos_leader(self,
+                                                            master_ar_process,
+                                                            valid_user_header):
+
+        # FIXME: using MesosDNS `leader.mesos` alias makes things hard to test.
+        # Dropping in in favour of cache+API call would improve reliability as
+        # well. So no "changing the leader and testing results tests for now"
+        generic_correct_upstream_dest_test(master_ar_process,
+                                           valid_user_header,
+                                           '/system/v1/leader/mesos/foo/bar',
+                                           'http://127.0.0.2:80',
+                                           )
+
+    def test_if_request_is_sent_to_the_current_marathon_leader(
+            self, master_ar_process, valid_user_header):
+
+        generic_correct_upstream_dest_test(master_ar_process,
+                                           valid_user_header,
+                                           '/system/v1/leader/marathon/foo/bar',
+                                           'http://127.0.0.2:80',
+                                           )
+
+        # Changing leader is covered in cache tests
+
+    @pytest.mark.parametrize("endpoint_type", [
+        ("marathon"),
+        ("mesos"),
+    ])
+    @pytest.mark.parametrize("sent,expected", [
+        ('/foo/bar?key=value&var=num',
+         '/foo/bar?key=value&var=num'),
+        ('/foo/bar/baz',
+         '/foo/bar/baz'),
+        ('/',
+         '/'),
+        ('',
+         ''),
+    ])
+    def test_if_upstream_request_is_correct(self,
+                                            master_ar_process,
+                                            valid_user_header,
+                                            sent,
+                                            expected,
+                                            endpoint_type):
+
```

Comment:
It would be wonderfull to have this as a flake8 rule/check - we could agree on something and have it enforced across all repos.

Fixed.